### ISSUE TITLE
Rule proposal: report unnecessarily parenthesised unions/intersections

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#6933]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

It's possible to unnecessarily add parentheses to unions/intersections. These parentheses are useless and provide no value to the code. This is a stylistic/formatting problem - and I don't usually report or support formatting rules, but this one is a *little* different.

Currently formatters do not touch the suggested fail cases:
- [prettier](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAKtgXmwAoBGbAH2wCYBKS7AZgG4AdWTHAVUJPKrrYAZEzYcs2AGq8ywmvREt2IADQgIGGAEtoAZ2SgAhgCdjEAO4AFEwn0pDAG3OG0+tQCNjhsAGs4MAGVDAFs4ABktKDhkADNHXTgPL19-AIxvSIBzZBhjAFdEkDhg9zgAEzLysMMoTLzDTLgAMQhjYMMYbVrkEEM8mAhVEAALGGCHAHVhrXhddLA4ANsZrQA3GbQesF03EEiE4xhLL0z22PjCgCtdAA8ArIc4AEU8iHhzhwS1dOMDnvQsLowMYtJohhgQbAJloyjBhsgABwABm+ZgSEy8GB6ELgB1W0TUAEdXvBjho7L1dABaKLlcpDYxwYlaRnHBpnJBxT6FBLBLQ5fI8h5wACCnRB7n6cEscGMESiHy+IF0wpeb2inIuahghnc0Nh8KQ1G1Xi0DiyAGEIMEOUVdABWIZ5BK4XV2LlK1YFACSUEqsACwNBMBFfoC6EeirgAF9o0A) (https://github.com/prettier/prettier/issues/13500)
- [rome](https://docs.rome.tools/playground/?code=dAB5AHAAZQAgAFQAIAA9ACAAKAAxACAAfAAgADIAKQAgAHwAIAAzADsACgB0AHkAcABlACAAVQAgAD0AIAAoADEAIAB8ACAAMgApACAAJgAgADMAOwAKAHQAeQBwAGUAIABWACAAPQAgACgAMQAgACYAIAAyACkAIAAmACAAMwA7AAoA) (https://github.com/rome/tools/issues/4388)
- [dprint](https://dprint.dev/playground/#code/C4TwDgpgBAKlC8UAUBGKAfKAmAlBqAzANwCwAUKJFAKoLJqa5QBkhpF40Aanai9nlbFyQA/language/typescript)
    - dprint is a bit different because dprint will intentionally leave in unnecessary parens in a lot of cases - it's inconsitent like that because it likes to assume the dev put the parens in for a reason (usually they didn't though, lol).

Prettier is open to fixing it, but there's definitely scope for a rule to help standardise codebases on older versions of prettier, and those without a formatter.

One could argue that you could [implement this solely using `no-restricted-syntax`](https://typescript-eslint.io/play/#ts=5.0.3&sourceType=module&showAST=es&code=PQKgsAUABFCmDOAbAlgOwC5VQewLQCcF19kBjdWAE13gE8MBDADwC4oBtSGGAcln3zZ8PADRduAb3HcoPeLESxyQnm1UBbBulIALBAAoAKgGUAqqmTZUh2gAdYUAHxQT5y9buwRL4wEkM-PLk7jb2Tj7+FPhB6CGeAJSi0tw86gjwDADmsKqy5gDu+Ay2UOh6ULYMhBh68vA8yQC+4gC6kCDAkOieLlAAvFD6AIxQAD5QAEzxY1AAzADcXT2m-YMj41NQAGRzixDdYQBqq8Pbk9M7C5BAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) - however I think there is additional value in a custom rule here so that it can have an auto-fixer to remove the parentheses. This would be good as in turn [it would then allow formatters to more consistently format the code](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACA7gSxgCwwFcoo4w4BnSgQwCcBPDAB3oQKqoyQAoAdWA2ZwMAFQwBeDD2ACMGGkgyUYdHFADmGAL4YAPhmAYARkpVrNOgJT7DGMGdXqt2gQPTY8+CERjFS5FS0jCxssPiclNxWAjBCIuJSslDyispOlroGRqbpFi62Rg55zjogADQgEMwwONCUyKD0dBBYAAphDSg0ADZYNAwNlcZ0NGAA1nAwAMo0ALZwADLqcMgAZr2UcMOjE1PTrGDOyKpE2yBwc8ZwACY3t4s0mkQ0GnAAYhB0czQwtZrIEA0XwQCogfAwOY9ADq+DwVEOcGmCEoeBwADc8AxAWBqGD1Fs6DA2qMND91ptzgArSgAD2mzh6cAAikQIPAKT0tpVWHRCYC4sJKGA1DUwcwLDBoTgbgRkAAOAAMPJaW2ho2YgIlnDo6NWlQAjmz4CTql0gZQALRkW63MF0OBGnAOkmvclIDZc85bOY4E50M6VVGaJkAQT+amMvjgbTgdGWZE53JAwY0TNZ7NWHsplRgNGM0tl+GQACZc6McD1nABhCBzd0XSgAVjBRC2onzXU9yfRZwAklB7rBpiKcDVQ4PpnEmUm4NptEA).

### Fail Cases

```typescript
type T1 = (1 | 2) | 3;
type T2 = (1 & 2) & 3;
```

### Pass Cases

```typescript
type T1 = 1 | 2 | 3;
type T2 = 1 & 2 & 3;
type T3 = (1 | 2) & 3;
type T4 = (1 | 2) & (3 | 4);
```

**Comment by bradzacher** [comment]
@JoshuaKGoldberg don't hate me pls sir I beg for forgiveness when zalgo comes

**Comment by Josh-Cena** [comment]
Has this been proposed in ESLint core (and would they accept such a rule)? My idea would be that a core rule called `no-unnecessary-parens` exists first that handles binary operators like `2 ** (3 ** 4)` and `2 * (3 * 4)`, and then we extend it to handle `2 | (3 | 4)`.

Another question is whether we support other kinds of unnecessary parens, like `keyof (typeof X)` or `(typeof X)[K]`, which are already handled by Prettier.

**Comment by bradzacher** [comment]
There is https://eslint.org/docs/latest/rules/no-extra-parens and we have an extension for it https://typescript-eslint.io/rules/no-extra-parens/

The issue I have is that turning on the rule might conflict with formatters because it impacts so much. A new rule could limit the scope to just this one issue that we know (currently) is not handled by formatters and thus won't conflict.

**Comment by Josh-Cena** [comment]
Might want to add more granular options? This rule would be so hard to name...

**Comment by bradzacher** [comment]
it does have some options - but I kind of feel like we'd want to add a new option like `types-only` - and then the issue there is that we would need to build it to handle every single types case. It'd be hard to justify doing this work in the extension rule for *just* the union/intersection case

Note that prettier does handle all the other cases because other cases don't change the AST it's just the `|``&` case.

**Comment by JoshuaKGoldberg** [comment]
...if you can have this rule, can I have #6494? 😄 

Feels like we should wait on a resolution to https://github.com/prettier/prettier/issues/13500 before deciding on this one, yeah?

**Comment by JoshuaKGoldberg** [comment]
Marking this as blocked on https://github.com/prettier/prettier/issues/13500 to remove it from my queue - if that stalls forever I suppose we could always take a look.

**Comment by bradzacher** [comment]
Hi there 👋!

As you may have heard [ESLint core is deprecating its formatting rules](https://eslint.org/blog/2023/10/deprecating-formatting-rules/). We are likewise moving in the same direction and deprecating our formatting rules.

Deprecated rules will be removed in the next major version as their new home is the [eslint-stylistic](https://github.com/eslint-stylistic/eslint-stylistic) project.

If this issue is important to you please consider filling an issue over at eslint-stylistic which will be the new home for stylistic/formatting rules.

Thanks!

**Comment by bradzacher** [comment]
This was my own issue but closing anyways because that's how much I hate formatting rules
